File: CaptureLive/App/helpers/ui/style/UIFont+CM.m
//
//  UIFont+CapturePresets.m
//  Capture
//
//  Created by Max Mikheyenko on 7/22/13.
//  Copyright (c) 2013 Capture. All rights reserved.
//

#import "UIFont+CM.h"
#import <CoreText/CoreText.h>

@implementation UIFont (CapturePresets)

+ (void)loadAssetFonts {
    [self loadAssetFontWithName:@"Comfortaa-Bold"];
    [self loadAssetFontWithName:@"Comfortaa-Light"];
    [self loadAssetFontWithName:@"Comfortaa-Regular"];
    [self loadAssetFontWithName:@"SourceSansPro-Black"];
    [self loadAssetFontWithName:@"SourceSansPro-BlackItalic"];
    [self loadAssetFontWithName:@"SourceSansPro-Bold"];
    [self loadAssetFontWithName:@"SourceSansPro-BoldItalic"];
    [self loadAssetFontWithName:@"SourceSansPro-ExtraLight"];
    [self loadAssetFontWithName:@"SourceSansPro-ExtraLightItalic"];
    [self loadAssetFontWithName:@"SourceSansPro-Italic"];
    [self loadAssetFontWithName:@"SourceSansPro-Light"];
    [self loadAssetFontWithName:@"SourceSansPro-LightItalic"];
    [self loadAssetFontWithName:@"SourceSansPro-Regular"];
    [self loadAssetFontWithName:@"SourceSansPro-Semibold"];
    [self loadAssetFontWithName:@"SourceSansPro-SemiboldItalic"];
}

+ (void)loadAssetFontWithName:(NSString *)name {
//    dumpAllFonts();
    NSString *fontPath                                          = [[NSBundle mainBundle] pathForResource:name ofType:@"ttf"];
    NSData *inData                                              = [NSData dataWithContentsOfFile:fontPath];
    CFErrorRef error;
    CGDataProviderRef provider                                  = CGDataProviderCreateWithCFData((__bridge CFDataRef)inData);
    CGFontRef font                                              = CGFontCreateWithDataProvider(provider);
    
    if (!CTFontManagerRegisterGraphicsFont(font, &error)) {
        CFStringRef errorDescription                            = CFErrorCopyDescription(error);
        NSLog(@"Failed to load font: %@", errorDescription);
        CFRelease(errorDescription);
    }
    CFRelease(font);
    CFRelease(provider);
}

//static void dumpAllFonts() {
//    for (NSString *familyName in [UIFont familyNames]) {
//        for (NSString *fontName in [UIFont fontNamesForFamilyName:familyName]) {
//            NSLog(@"%@", fontName);
//        }
//    }
//}


@end

File: CaptureLive/App/helpers/ui/style/UIFont+CM.swift
//
//  UIFont+CM.swift
//  Capture-Live
//
//  Created by hatebyte on 7/7/15.
//  Copyright © 2015 CaptureMedia. All rights reserved.
//

import UIKit

public enum CMFont : String {
    case Thin                       = "ProximaNova-Thin"
    case Light                      = "ProximaNova-Light"
    case Regular                    = "ProximaNova-Regular"
    case Bold                       = "ProximaNova-Bold"
}

public enum CMFontSize : CGFloat {
    case XSmall                     = 10
    case Small                      = 12
    case Medium                     = 15
    case Large                      = 18
    case XLarge                     = 22
    case XLarge28                   = 28
    case XXLarge                    = 32
}

public enum CMUIFontSize : CGFloat {
    case NavigationTitle            = 16
}

public enum CMFontComfortaa : String {
    case Light                      = "Comfortaa-Light"
    case Regular                    = "Comfortaa-Regular"
    case Bold                       = "Comfortaa-Bold"
}

public enum CMFontSourceSansPro : String {
    case Bold                       = "SourceSansPro-Bold"
    case BoldItalic                 = "Source-Sans-Pro-BoldItalic"
    case ExtraLight                 = "SourceSansPro-ExtraLight"
    case ExtraLightItalic           = "SourceSansPro-ExtraLightIt"
    case Italic                     = "SourceSansPro-It"
    case Light                      = "SourceSansPro-Light"
    case LightItalic                = "SourceSansPro-LightIt"
    case Regular                    = "SourceSansPro-Regular"
    case SemiBold                   = "SourceSansPro-Semibold"
    case SemiBoldItalic             = "SourceSansPro-SemiboldIt"
}

extension UIFont {

    public class func comfortaa(comfortaa: CMFontComfortaa, size fontSize: CMFontSize)->UIFont {
        return UIFont(name: comfortaa.rawValue, size: fontSize.rawValue)!
    }
    
    public class func comfortaa(comfortaa: CMFontComfortaa, size fontSize: CMUIFontSize)->UIFont {
        return UIFont(name: comfortaa.rawValue, size: fontSize.rawValue)!
    }
    
    public class func sourceSansPro(sourceSansPro: CMFontSourceSansPro, size fontSize: CMFontSize)->UIFont {
        return UIFont(name: sourceSansPro.rawValue, size: fontSize.rawValue)!
    }
    
    public class func comfortaa(comfortaa: CMFontComfortaa, size fontSize: CGFloat)->UIFont {
        return UIFont(name: comfortaa.rawValue, size: fontSize)!
    }
    
    public class func sourceSansPro(sourceSansPro: CMFontSourceSansPro, size fontSize: CGFloat)->UIFont {
        return UIFont(name: sourceSansPro.rawValue, size: fontSize)!
    }

    public func sizeOfString(string: String, constrainedToWidth width: Double) -> CGSize {
        return NSString(string: string).boundingRectWithSize(CGSize(width: width, height: DBL_MAX),
            options: [.UsesLineFragmentOrigin, .UsesFontLeading],
            attributes: [NSFontAttributeName: self],
            context: nil).size
    }
    
}